pp_hot.c

```diff
@@ -4975,8 +4975,8 @@ Perl_clear_defarray(pTHX_ AV* av, bool abandon)
         AvREIFY_only(av);
     }
     else {
-        AV *newav = newAV();
-        av_extend(newav, fill);
+        /* fill can be -1. The ternary gives consistency with av_extend() */
+        AV *newav = newAV_alloc_xz(fill < 3 ? 4 : fill);
```

Comment:
This looks wrong, because `newAV_alloc_xz()` takes a size rather than a key - it should be `fill < 3 ? 4 : fill + 1`. For cleanliness, it would probably be better instead to set `const SSize_t size = AvFILLp(av) + 1; ... newAV_alloc_xz(size < 4 ? 4 : size)`

Note also that the original `av_extend()` call zero-filled the array only because the `AvREIFY_only()` is applied late; a small optimization therefore would be to use `newAV_alloc_x()` instead. (But since that's a change of behaviour, it's probably worth calling out with a separate commit.)

Reply:
Gah, you're right. Managed to get confused by key vs size myself.

The point about `AvREIFY_only()` is a good one. I'll double check: might well have left it out since it was a change, but half recollect trying to just call `newAV_alloc_x()` and being surprised when that broke things.

Reply:
I was thinking about some other refactoring breaking with `newAV_alloc_x()`. I've added the change to use that as a separate commit (to remain unsquashed at the end).